base/compiler/tfuncs.jl

```diff
@@ -1425,8 +1443,38 @@ function tuple_tfunc(argtypes::Vector{Any})
     return anyinfo ? PartialStruct(typ, argtypes) : typ
 end
 
-function arrayref_tfunc(@nospecialize(boundscheck), @nospecialize(a), @nospecialize i...)
-    a = widenconst(a)
+function arrayref_tfunc(@nospecialize(boundcheck), @nospecialize(ary),
+    @nospecialize idxs...)
+    isempty(idxs) && return Bottom
+    array_builtin_common_errorcheck(boundcheck, ary, idxs) || return Bottom
+    return array_elmtype(ary)
+end
+add_tfunc(arrayref, 3, INT_INF, arrayref_tfunc, 20)
+add_tfunc(const_arrayref, 3, INT_INF, arrayref_tfunc, 20)
+
+function arrayset_tfunc(@nospecialize(boundcheck), @nospecialize(ary), @nospecialize(item),
+    @nospecialize idxs...)
+    isempty(idxs) && return Bottom
+    array_builtin_common_errorcheck(boundcheck, ary, idxs) || return Bottom
+    hasintersect(widenconst(item), array_elmtype(ary)) || return Bottom
```

Comment:
```suggestion
    hasintersect(widenconst(item), arrayref_tfunc(boundcheck, ary, idxs...)) || return Bottom
```

  Also, why boundcheck instead of boundscheck?